### PR TITLE
Optimize FFT calculations

### PR DESCRIPTION
**Pull request by sultanqasim**
Use multiple threads for large FFT sizes, and volk-ify hot loops. Needs `fftwf_init_threads()` to be called before sigutils is used (perhaps from `suscan_sigutils_init()`).

### CODE REVIEW

**Comment by sultanqasim** [comment]
Time to calculate apply window function, perform FFT, then calculate real value. Benchmarked at 2 Msps, 60 fps, with various FFT sizes. Overlapping buffers for FFT size 65536 and up.

~~~
Volk + FFT threads:
Size		Time (us)
16384		170
32768		350
65536		500
131072		1000
262144		2300

FFT threads only:
Size		Time (us)
16384		250 (single thread)
32768		450 (multi-thread, 550 single thread)
65536		1000
131072		1600
262144		3600

Original:
Size		Time (us)
16384		200
32768		400
65536		1450
131072		2200
262144		4900
~~~

I'm not quite sure why just linking libfftw3f_threads made the FFTs a bit slower, even with no other code changes. However, I did test to see that multiple threads improved performance relative to a single thread for FFT sizes of 32768 and higher.

**Comment by BatchDrake** [comment]
Perhaps old FFT wisdom? Remove the wisdom file from ~/.suscan, open SigDigger, capture some samples, close SigDigger (gracefully), open again and profile.

**Comment by sultanqasim** [comment]
I tried clearing out wisdom multiple times when I was testing but still had the issue. Interestingly, when I tried changing the planner mode to `FFTW_PATIENT` or `FFTW_EXHAUSTVE`, FFTs became a lot slower instead of faster (I'm not talking about planning time but actual FFT execution time).

**Comment by sultanqasim** [comment]
I accidentally benchmarked on a debug build earlier, so I did on a release build now. Actually, it seems the volk-ization was unnecessary, at least on my machine (ARM Mac). Seems the compiler was already SIMD-ifying these loops with -O3. Maybe Volk does better on x86, you can check. Othewise, the Volk change can be replaced with just that memcpy simplification I did for the overlapping buffers case.

~~~
Volk + FFT threads:
Size		Time (us)
8192		53
16384		115
32768		250
65536		570
131072		820
262144		2100

FFT threads only:
Size		Time (us)
8192		50
16384		110
32768		250
65536		550
131072		800
262144		2100

Original:
Size		Time (us)
8192		50
16384		105
32768		220
65536		750
131072		1200
262144		3100
~~~

Note: while FFTW3F runs clearly faster with multiple threads for size 65536, and slightly faster with multiple threads for size 32768, on my computer in my experiments with libfftw3f_threads linked, the linking of libfftw3f_threads has the strange side-effect of adding a lot of jitter to the computation time of FFTs. For example, the 8192 size FFTs sometimes take 50ms, and other times 150+ ms. I wonder if libfftw3f_threads is causing the FFT to be done in a separate worker thread even if the planner is asked to only use one thread.

**Comment by sultanqasim** [comment]
While it seems using Volk actually slightly worsens performance on release builds in my M1 MacBook Air, this one subset of the volk-ization commit did improve performance as expected: https://github.com/sultanqasim/sigutils/commit/eac9bf5e88a2bd1b20c69b2ac0f2014b96c3e21c

If Volk doesn't help on your machine either, then perhaps we should not use Volk for these calculations, but still do that memcpy optimization to avoid branching inside the loop.

**Comment by sultanqasim** [comment]
I added proper locking to protect the non-thread safe fftw planner. This seems to fix the jitter issues I was having with single threaded work still happening in another thread. I now migrated all the FFT operations to use this multi-threaded and lock protected planner. I use two threads instead of four for 32k FFT size, as the performance benefit of four threads is negligible on my machine, but four threads does result in a lot more jitter and inefficiency at 32k size.

**Comment by sultanqasim** [comment]
The last commit of this series (the volkization) is completely useless on my M1 MacBook Air, actually it slightly worsens performance. I don't have an x86 machine handy to test if there's any benefit there, but assuming not, that last volk-ization commit should be discarded.

**Comment by sultanqasim** [comment]
Got rid of the volk commit for now, and switched to FFTW's built in thread safety mechanism. Working great for me.

**Comment by BatchDrake** [comment]
Thanks! Testing it now

File: src/sigutils/lib.c [hunk 1]
```diff
@@ -126,6 +128,30 @@ su_lib_fftw_strategy(void)
...
+    nthreads = 2;
+  else
+    nthreads = 4;
+
+  if (pthread_mutex_lock(&fft_plan_mutex) == -1)
+    return NULL;
+  SU_FFTW(_plan_with_nthreads)(nthreads);
+  plan = SU_FFTW(_plan_dft_1d)(n, in, out, sign, flags);
+  SU_FFTW(_plan_with_nthreads)(1);
+  pthread_mutex_unlock(&fft_plan_mutex);
```
Comment (BatchDrake) [inline src/sigutils/lib.c:150]: Next time consider prefixing global variables by `g_`. While there are still a bunch of variables that do not follow this convention, it helps me a lot when I am reading new code.

File: src/sigutils/lib.c [hunk 2]
```diff
@@ -126,6 +128,30 @@ su_lib_fftw_strategy(void)
...
+{
+  SU_FFTW(_plan) plan;
+  int nthreads;
+
+  if (n < 32768)
+    nthreads = 1;
+  else if (n == 32768)
+    nthreads = 2;
+  else
+    nthreads = 4;
```
Comment (BatchDrake) [inline src/sigutils/lib.c:143]: I am going to mere this as-is. But perhaps, as an improvement for the future, we can choose the number of threads depending on the number of available cores in a system.

**Pull request merged**